File: posts/bench/bench.py
import timeit
import random


class O:
    def __init__(self):
        self.x = random.randint(0, 100)
        self.y = random.randint(0, 100)
        self.z = random.randint(0, 100)
        self.should_foo = random.choice([True, False])


class O2:
    __slots__ = ('x', 'y', 'z', 'should_foo')

    def __init__(self):
        self.x = random.randint(0, 100)
        self.y = random.randint(0, 100)
        self.z = random.randint(0, 100)
        self.should_foo = random.choice([True, False])


def new_a():
    return [
        random.randint(0, 100),
        random.randint(0, 100),
        random.randint(0, 100),
        random.choice([True, False])
    ]

def new_t():
    return (
        random.randint(0, 100),
        random.randint(0, 100),
        random.randint(0, 100),
        random.choice([True, False])
    )


Os = [O() for _ in range(10000)]
O2s = [O2() for _ in range(10000)]
As = [new_a() for _ in range(10000)]
Ts = [new_t() for _ in range(10000)]


def as_object():
    acc = 0
    acc2 = 0

    for o in Os:
        val = (o.x + o.y) * o.z
        acc += val
        if o.should_foo:
            acc2 += val


def as_optimized_object():
    acc = 0
    acc2 = 0

    for o in O2s:
        val = (o.x + o.y) * o.z
        acc += val
        if o.should_foo:
            acc2 += val


def as_list():
    acc = 0
    acc2 = 0

    for a in As:
        val = (a[0] + a[1]) * a[2]
        acc += val
        if a[3]:
            acc2 += val

def as_tuple():
    acc = 0
    acc2 = 0

    for t in Ts:
        val = (t[0] + t[1]) * t[2]
        acc += val
        if t[3]:
            acc2 += val


def bench(f):
    time = timeit.timeit(
        "{}()".format(f),
        setup="from __main__ import {}".format(f),
        number=10000)

    print("{}: {}".format(f, time))


if __name__ == "__main__":
    bench("as_object")
    bench("as_list")
    bench("as_optimized_object")

File: posts/bench/bench.nim
import nimbench, random

randomize()

type
  O = object
    x, y, z: int
    shouldFoo: bool
  A = array[0..3, int]

var
  Os = newSeq[O](1_000_000)
  As = newSeq[A](1_000_000)

for i in Os.low..Os.high:
  let
    x = rand(100)
    y = rand(100)
    z = rand(100)
    shouldFoo = rand(1)
  Os[i] = O(
    x: x,
    y: y,
    z: z,
    shouldFoo: shouldFoo.bool
  )
  As[i] = [x, y, z, shouldFoo]

bench(asObject, m):
  var acc, acc2: int
  for _ in 1..m:
    for o in Os:
      let val = (o.x + o.y) * o.z

      acc += val
      if o.shouldFoo:
        acc2 += val

  doNotOptimizeAway(acc)
  doNotOptimizeAway(acc2)

bench(asArray, m):
  var acc, acc2: int
  for _ in 1..m:
    for a in As:
      let val = (a[0] + a[1]) * a[2]

      acc += val
      if a[3] > 0:
        acc2 += val

  doNotOptimizeAway(acc)
  doNotOptimizeAway(acc2)

runBenchmarks()
